compiler/rustc_const_eval/src/interpret/eval_context.rs

```diff
@@ -1069,23 +1070,33 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         })
     }
 
-    pub fn eval_global(
+    pub fn eval_global_scalar(
         &self,
         instance: ty::Instance<'tcx>,
-    ) -> InterpResult<'tcx, MPlaceTy<'tcx, M::Provenance>> {
+    ) -> InterpResult<'tcx, Scalar<M::Provenance>> {
         let gid = GlobalId { instance, promoted: None };
-        // For statics we pick `ParamEnv::reveal_all`, because statics don't have generics
-        // and thus don't care about the parameter environment. While we could just use
-        // `self.param_env`, that would mean we invoke the query to evaluate the static
-        // with different parameter environments, thus causing the static to be evaluated
-        // multiple times.
-        let param_env = if self.tcx.is_static(gid.instance.def_id()) {
-            ty::ParamEnv::reveal_all()
+        let (alloc_id, alloc) = if self.tcx.is_static(gid.instance.def_id()) {
+            (
+                self.tcx.reserve_and_set_static_alloc(gid.instance.def_id()),
+                self.ctfe_query(|tcx| tcx.eval_static_initializer(gid.instance.def_id()))?,
+            )
         } else {
-            self.param_env
+            let val = self.ctfe_query(|tcx| tcx.eval_to_allocation_raw(self.param_env.and(gid)))?;
+            (val.alloc_id, self.tcx.global_alloc(val.alloc_id).unwrap_memory())
         };
-        let val = self.ctfe_query(|tcx| tcx.eval_to_allocation_raw(param_env.and(gid)))?;
-        self.raw_const_to_mplace(val)
+
+        let ty = instance.ty(self.tcx.tcx, self.param_env);
+        let layout = self.layout_of(ty)?;
+        let read_provenance = matches!(
+            layout.abi,
+            Abi::Scalar(abi::Scalar::Initialized { value: abi::Primitive::Pointer(..), .. })
+        );
+
+        let scalar = alloc
+            .inner()
+            .read_scalar(self, alloc_range(Size::ZERO, layout.size), read_provenance)
+            .map_err(|err| err.to_interp_error(alloc_id))?;
+        self.adjust_const_scalar(scalar)
```

Comment:
As before I'm not happy with bypassing all our nice abstractions here. Why is this necessary?

Reply:
I got rid of it (see the last commit for how it works now)